# 批量处理流程控制及全程接口异常处理

为了进行文件的复制，移动，删除，查毒等操作，后端给了一套批量处理流程的接口。其流程是这样的：

1. 调用 `createTask` 接口创建批量处理任务，获取 taskId
2. 根据 taskId 轮询 `checkTask` 接口查询当前任务的处理情况（处理完成或是处理中）
3. 点击“取消”按钮则调用 `cancelTask` 接口结束任务。

在接口返回时调用其他接口也能实现需求，但这种回调地狱的写法不是我想要的。再次分析需求，可以发现这个流程是符合有限状态机的模型的。当前的任务状态只能处于初始状态，创建任务中，查询任务中，取消任务中，任务结束和任务异常（接口异常进入异常状态）这几种状态。且同一时间只能处于一种状态，状态之间可以转换。

除此之外，状态的转换一般是在接口返回的时候，因此，我们要在接口返回的时候检测状态，根据状态决定下一步流程。如果是这样的话，其实还是在回调中处理下一步。因此，我换了一个思路，轮询状态，根据不同状态决定下一步操作。每一个接口返回时更新state的值以便下一个循环检查。

```js
// state: init create check cancel finish error
let state = 'init'
const task = async (params) => {
    state = 'create'
    initData(params) // 数据初始化
    while (state !== 'finish' && state !== 'error') {
        switch (state) {
        case 'create':
            await createTask()
            break
        case 'check':
            await sleep(1000)
            await checkTask()
            break
        case 'cancel':
            await cancelTask()
            break
        }
    }
    // 成功 / 异常 提示
    showDefaultMessage()
}
```

接下来是每个接口的异常捕获，每个接口异常时将 state 设置成 error 即可退出循环。实际业务中，一个操作可能需要进行两个task（如复制时对文件进行查毒扫描），所以 task 也需要返回一个 Promise。那么在接口异常的时候，就需要将 task 返回的 Promise 的状态设置成 reject。即在每个接口捕获异常的时候，调用 task Promise 的 reject 方法。因此封装了一个异步方法来调用这些接口。

```js
const asyncFunc = (func, promise) => {
    return func()
        .then(res => {
            (state === 'finish') && promise.resolve(res)
        })
        .catch(err => {
            state = 'error'
            promise.reject(err)
        })
}

let state = 'init'
const task = async (params) => {
    state = 'create'
    initData(params) // 数据初始化
    return new Promise(async (resolve, reject) => {
        let promise = { resolve, reject }
        while (state !== 'finish' && state !== 'error') {
            switch (state) {
            case 'create':
                await asyncFunc(createTask, promise)
                break
            case 'check':
                await sleep(1000)
                await asyncFunc(checkTask, promise)
                break
            case 'cancel':
                await asyncFunc(cancelTask, promise)
                break
            }
        }
        // 成功 / 异常 提示
        showDefaultMessage()
    })
}

// 调用
task(params).then(res => {
    // 任务完成
}).catch(err => {
    // 任务中某个接口异常或任务取消
})
```

取消任务还有一些问题。需求中，只能在 check 的时候取消任务。取消任务是事件触发的，触发后 state 设置为 cancel，此时 check 接口还没返回，这会导致 check 接口返回时修改 state 的状态，导致无法取消。解决方法为取消任务前先取消 check 接口请求。这样又引发了另外的问题：中断请求导致整个流程进入了 error，同时后续的 cancelTask 也无法进入。

因此，对于一些特殊情况，需要对接口的 Promise 的状态进行特殊的设置。

```js
let checkCancelSource = null
const checkTask = () => {
    return new Promise((resolve, reject) => {
        let CancelToken = axios.CancelToken
        checkCancelSource = CancelToken.source() // 用于取消check请求
        let cancelToken = checkCancelSource.token
        axios({...params, cancelToken}).then(res => {
            ...
            resolve()
        }).catch(err => {
            // 如果是主动取消，则 checkTask 的状态仍设置为 resolve，以进行下一步流程，避免进入 error
            axios.isCancel(err) ? resolve() : reject(err)
        })
    })
}

const cancelTask = () => {
    return new Promise((resolve, reject) => {
        axios(params).then(res => {
            // 对于 task 来说，任务取消应该进入 catch
            // 所以此处使用 reject
            reject(new Error('任务取消'))
        }).catch(reject)
    })
}
```

实际业务中，还需要处理文件的冲突。当 check 接口返回文件冲突的状态码时，需要调用接口获取冲突的文件，用户选择保留、跳过或者替换操作。

这只需要在上面的流程加入两个状态即可（获取冲突文件、处理冲突）。而主要的难点在于处理冲突需要等待用户操作，我们需要用 await 等待用户输入，再进行下一步，即把事件回调函数变成 async / await 这样的同步写法。

```js
const task = async (params) => {
    // ...
    return new Promise(async (resolve, reject) => {
        let promise = { resolve, reject }
        while (state !== 'finish' && state !== 'error') {
            switch (state) {
            // ...
            case 'conflict':
                await asyncFunc(getConflict, promise)
                await conflictHandle() // 用户处理冲突
                break
            }
        }
        // 成功 / 异常 提示
        showDefaultMessage()
    })
}

let conflictResolve = () => {}
let conflictReject = () => {}
// 创建一个未变状态的 Promise
// 根据用户操作去改变其状态
const conflictHandle = async () => {
    try {
        await (() => new Promise((resolve, reject) => {
            conflictResolve = resolve
            conflictReject = reject
        }))()
        state = 'check'
    } catch {
        state = 'cancel'
    }
}
// 用户选择了处理方式
const confirmHandle = (type) => {
    // deal type
    conflictResolve()
}
// 用户选择关闭，取消任务
const cancelHandle = () => {
    conflictReject()
}
```

这个方案的主要优势是将多个异步流程改造成同步写法，避免了回调地狱。且进行了整个流程的接口异常处理。并将多个异步流程封装成一个对外的异步接口，让调用更方便。